src/main/java/edu/harvard/iq/dataverse/search/IndexServiceBean.java

```diff
@@ -348,48 +349,78 @@ public Future<String> indexDataverse(Dataverse dataverse, boolean processPaths)
+    public void indexDatasetInNewTransaction(Long datasetId) { //Dataset dataset) {
         boolean doNormalSolrDocCleanUp = false;
         Dataset dataset = em.find(Dataset.class, datasetId);
-        // return indexDataset(dataset, doNormalSolrDocCleanUp);
-        Future<String> ret = indexDataset(dataset, doNormalSolrDocCleanUp);
+        asyncIndexDataset(dataset, doNormalSolrDocCleanUp);
         dataset = null;
-        return ret;
     }
+
+    // The following two variables are only used in the synchronized getNextToIndex method and do not need to be synchronized themselves
     
-    @TransactionAttribute(REQUIRES_NEW)
-    public Future<String> indexDatasetObjectInNewTransaction(Dataset dataset) throws  SolrServerException, IOException{ //Dataset dataset) {
-        boolean doNormalSolrDocCleanUp = false;
-        // return indexDataset(dataset, doNormalSolrDocCleanUp);
-        Future<String> ret = indexDataset(dataset, doNormalSolrDocCleanUp);
-        dataset = null;
-        return ret;
+    // nextToIndex contains datasets mapped by dataset id that were added for future indexing while the indexing was already ongoing for a given dataset
+    // (if there already was a dataset scheduled for indexing, it is overwritten and only the most recently requested version is kept in the map)
+    private static Map<Long, Dataset> nextToIndex = new HashMap<>();
+    // indexingNow is a set of dataset ids of datasets being indexed asynchronously right now
+    private static Set<Long> indexingNow = new HashSet<>();
+
+    // When you pass null as Dataset parameter to this method, it indicates that the indexing of the dataset with "id" has finished
+    // Pass non-null Dataset to schedule it for indexing
+    synchronized private static Dataset getNextToIndex(Long id, Dataset d) {
+        if (d == null) { // -> indexing of the dataset with id has finished
+            Dataset next = nextToIndex.remove(id);
+            if (next == null) { // -> no new indexing jobs were requested while indexing was ongoing
+                // the job can be stopped now
+                indexingNow.remove(id);
+            }
+            return next;
+        }
+        // index job is requested for a non-null dataset
+        if (indexingNow.contains(id)) { // -> indexing job is already ongoing, and a new job should not be started by the current thread -> return null
+            nextToIndex.put(id, d);
+            return null;
+        }
+        // otherwise, start a new job
+        indexingNow.add(id);
+        return d;
     }
 
     @Asynchronous
-    public Future<String> asyncIndexDataset(Dataset dataset, boolean doNormalSolrDocCleanUp) throws  SolrServerException, IOException {
-        return indexDataset(dataset, doNormalSolrDocCleanUp);
+    public void asyncIndexDataset(Dataset dataset, boolean doNormalSolrDocCleanUpe) {
```

Comment:
I completely agree with you that, as we ignore the returned `Future` anyway everywhere in the code, we can make this `void`. Yet, I'd argue that we could take this even a step further.

Instead of using `@Asynchronous`, an EJB specific mechanism to execute this method decoupled from the rest of the calling code, maybe it would be a good time to start using an [`Executor`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/Executor.html). We could deploy tasks/datasets to a queue (both have implementations in `java.util.concurrent`), making an executor pick them up. This would even enable scaling the indexing onto multiple threads, as executors can be pooled.

Using a `Queue` also enables adding metrics for the status and throughput via MicroProfile Metrics, so people can adapt the queue size and executor pools (and all that comes with that like resources etc).

Reply:
From what I understand, calls to @Asynchronous methods are executed with the EJB thread pool, which has by default 16 core threads. These are then managed by admins: https://docs.payara.fish/enterprise/docs/Technical%20Documentation/Payara%20Server%20Documentation/Server%20Configuration%20And%20Management/Thread%20Pools/EJB%20Thread%20Pool.html

I am not sure if it is a good idea of creating extra thread-pools (or does it use the same thread EJB thread pool)? Maybe if we want to make a more limited thread-pool with only few threads? But on the other hand, we want the indexing to finish as soon as possible, since some code might be waiting for the results. Nevertheless, it is something worth to consider as it might have some benefits.

Reply:
Obviously one should use the Jakarta EE Concurrency `Executors`, because we's want the application server to manage the pool. Here's the spec for Java EE 8 https://jakarta.ee/specifications/concurrency/1.1 (note that mostly the Javadoc is useful there, for the docs better look into https://jakarta.ee/specifications/concurrency/2.0/concurrency-spec-2.0.html which is Jakarta EE 9, the renamed EE 8)

IMHO your implementation with the set is kind of like a queue already, but with one important difference: there are no mechanisms in place that avoid saturating the consumer.

A) If you just put in enough requests (for different datasets), the queue grows infinite. This might be OK for us, but the Solr service usually has resource limitations as well, so this might be something a sysadmin wants to tune for.

b) If your indexing is fast enough, it will not avoid double work. If reindexing takes 1 second and the same dataset comes in again after 3 seconds, it will still do two indexing ops. Some might say this is what we want, but on the other hand in large installations you might want to have tunable delay. It's probably OK to have an updated search index after some X seconds when edits come in hot (machine edits etc), but searches are less frequent.

C) There is no feedback to a user when indexing performance degrades. As the command engine is simply dropping of the workload in an infinite queue, we cannot provide a feedback response to the command engine that now is not a good time so please come back later. Such feedback should not result in aborting publications (as the indexing happens after all that done), but at least it should create log messages about performance degration or even notifications to the sysadmin. Then an admin might investigate and tune configuration and resources or schedule a reindex for off-hours to spread the load.

Reply:
I am not an expert in EJB, there might be a solution that is beyond my reach. I think I will not be able to implement it by myself.

I will give it my best shot and finish what I am doing now. We can decide what to do with that later on.

Reply:
I have moved the implementation to a clustered singleton bean with blocking thread pool executor, where the maximum number of threads is configured via microprofile. I have kept the skipping indexing logic and added more documentation about it. I think we are getting there.

I am not sure about the way that the executor is initialized, should it come from some managed executor factory or something? One that I checked was from the glassfish api and used exactly the same code that I used for its implementation.

It would be also very useful if I could see why the integration test are failing, running them by myself is very time-consuming. Maybe there could be something added that e-mails the errors automatically when the tests fail? I will rerun them later locally.

Reply:
I had to remove the blocking thread pool executor from the commit
https://github.com/IQSS/dataverse/pull/9558/commits/84116dd30351a4fc4f9bed2e882dafcfc016ac25

I noticed that when 4 users would be uploading large datasets simultaneously, the dataverse could be blocked for days (not allowing any updates to datasets), I went back to the previous solution with @Asynchronous annotation. It could degrade the performance when more users upload large datasets (containing thousands of files), but it should not block, as long as the payara has a decent task scheduling algorithm. I have also noticed that 26 other methods also use the @Asynchronous annotation, the problem could be bigger than just the indexing... I will do some more testing and simulate more users uploading large datasets.

Reply:
It looks like the @Asynchronous annotation has the same default behavior as the blocking thread pool executor, it blocks when the thread pool is exhausted. Luckily I was wrong, and it does not block for days, as new indexing is not accepted for any dataset until a thread becomes available. I will need some more testing to confirm that for sure, but I notice the wait when uploading multiple datasets simultaneously.

Reply:
I could not make it work, I gave up and reverted to the previous solution.